### ISSUE TITLE
A Question about the Calculation of cauchy_cross_entroy Loss 

### ISSUE THREAD

**Issue by MaxChanger** [thulab/DeepHash#28]
Hello, I recently looked at the DCH method in the DeepHash library, but I encountered some incomprehension, as shown in the following figure

![20190512213035](https://user-images.githubusercontent.com/35401968/57582945-b09ae580-74fd-11e9-8601-ffb08359619a.png)

The parameter `v` and `label_v` in function `cauchy_cross_entroy` is always` None`, so in function  `cauchy_cross_entroy`, the part in the red box is always executed. So u and v are the same.

At the same time, I'm also looking at HashNet's pytorch code implementation. And It also requires two sets to calculate the loss, but outputs1 and outputs2 in the figure below take two batches from the entire data, which **are different**.

![20190512213723](https://user-images.githubusercontent.com/35401968/57582973-27d07980-74fe-11e9-9d14-ce3f1d0600c2.png)

So now I'm wondering why you set U and V to the same directly.
In the paper, h_i and h_j are also used to represent two sets, which are not the same.

I hope I can get your answer. Thank you.

**Comment by bl0** [comment]
Hi, I think that calculating the loss within the batch and cross two batches are both ok. 
But obviously within the batch is more convenient.

**Comment by MaxChanger** [comment]
I found this problem when I tried to modify the DCH based on the HashNet code, because I was more familiar with pytorch. I modified the loss function of HashNet according to the DCH paper and tensorflow code, but if I used the image data from two batches to calculate loss, the result would be `nan`. But if the settings in the picture above come from the same batch, I can get a map on coco that is similar to the one in the paper, 0.73.

**Comment by MaxChanger** [comment]
I can't understand why it's set to come from the same batch.
I think if u and v come from the same batch, the values of h_i and h_j should be equal, h_i = h_j.
And if it's true, then dist(h_i,h_j) = dist(h_i,h_i) , what's the mean about dist(h_i,h_i) 
I don't know if I understand it correctly.

**Comment by bl0** [comment]
Hi, a batch also contains some samples. So i and j can be any number in [0, batch_size) and we can calculate the cross entropy loss for all pair (i, j) within the batch.

**Comment by bl0** [comment]
PYI: In our implementation, ip_1 and label_ip both have shape [bs, bs] where bs is the batch size.

**Comment by MaxChanger** [comment]
Hi, I think the `u` in the code is batch_size×hash_bits, for example 128×48, represent the batch has 128 images and the length of hash code is 48, the `label_u` is batch_size×label_bits, the label_bits depend on the train.txt/database.txt. Hence, as you said, ip_1 and label_ip both have shape [batch_size, batch_size].

But I understand that h_i or h_j is a row of `u` matrices, and if `ip_1= tf.matmul(u, tf.transpose(v)` represent matrix u is multiplied by v's transposition. if u and v are the same, It is equivalent to  `ip_1 = tf.matmul(u, tf.transpose(u)`, the value of ip_1[i][j] represents the dot product of u[h_i] row and u^T[h_j] column, but is the value of ip[i][i] meaningful?

**Comment by bl0** [comment]
Why ip[i][i] is not meaningful? They come from different i and j.

**Comment by MaxChanger** [comment]
ip[i][j] is meaningful, they come from different i and j, but ip[i][i] is dist(h_i, h_i) not dist(h_i, h_j), it represents a hash code of an image dot product by itself

**Comment by bl0** [comment]
Ok, I understand your question now. 
Yes, ip[i, i] may not be meaningful. But the loss for pair (i, i) will be zero and don't affect the training. So we need not worry that.

**Comment by MaxChanger** [comment]
Thank you for your answer. Through the discussion with you, I am more convinced that my understanding is correct or not.
But I still need to find why I set two different batch in my code as input led to the loss is `nan`, but the same  batch can work perfectly.
If I have some progress, I will show in this issue again. Thanks

**Comment by bl0** [comment]
Ok. Welcome to discuss with me.
